examples/governance/apps/src/bin/publisher.rs

```diff
@@ -0,0 +1,175 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This application demonstrates how to send an off-chain proof request
+// to the Bonsai proving service and publish the received proofs directly
+// to your deployed app contract.
+
+use alloy_sol_types::{sol, SolInterface};
+use anyhow::{Context, Result};
+use clap::Parser;
+use ethers::prelude::*;
+use hex::decode;
+use methods::FINALIZE_VOTES_ELF;
+use risc0_ethereum_contracts::groth16;
+use risc0_zkvm::{default_prover, ExecutorEnv, ProverOpts, VerifierContext};
+
+sol! {
+    interface RiscZeroGovernor {
+        function verifyAndFinalizeVotes(bytes calldata seal, bytes calldata journal) public;
+    }
+}
+
+/// Wrapper of a `SignerMiddleware` client to send transactions to the given
+/// contract's `Address`.
+pub struct TxSender {
+    chain_id: u64,
+    client: SignerMiddleware<Provider<Http>, Wallet<k256::ecdsa::SigningKey>>,
+    contract: Address,
+}
+
+impl TxSender {
+    /// Creates a new `TxSender`.
+    pub fn new(chain_id: u64, rpc_url: &str, private_key: &str, contract: &str) -> Result<Self> {
+        let provider = Provider::<Http>::try_from(rpc_url)?;
+        let wallet: LocalWallet = private_key.parse::<LocalWallet>()?.with_chain_id(chain_id);
+        let client = SignerMiddleware::new(provider.clone(), wallet.clone());
+        let contract = contract.parse::<Address>()?;
+
+        Ok(TxSender {
+            chain_id,
+            client,
+            contract,
+        })
+    }
+
+    /// Send a transaction with the given calldata.
+    pub async fn send(&self, calldata: Vec<u8>) -> Result<Option<TransactionReceipt>> {
+        let tx = TransactionRequest::new()
+            .chain_id(self.chain_id)
+            .to(self.contract)
+            .from(self.client.address())
+            .data(calldata);
+
+        log::info!("Transaction request: {:?}", &tx);
+
+        let tx = self.client.send_transaction(tx, None).await?.await?;
+
+        log::info!("Transaction receipt: {:?}", &tx);
+
+        Ok(tx)
+    }
+}
+
+/// Arguments of the publisher CLI.
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct Args {
+    /// Ethereum chain ID
+    #[clap(long)]
+    chain_id: u64,
+
+    /// Ethereum Wallet Private Key
+    #[clap(long, env)]
+    eth_wallet_private_key: String,
+
+    /// Node RPC URL
+    #[clap(long)]
+    rpc_url: String,
+
+    /// Application's contract address on Ethereum
+    #[clap(long)]
+    contract: String,
+
+    /// The proposal ID (32 bytes, hex-encoded)
+    #[clap(long)]
+    proposal_id: String,
+
+    /// The votes data (hex-encoded, multiple of 100 bytes)
+    #[clap(long)]
+    votes_data: String,
```

Comment:
How are folks expected to construct the vote data? In an end-to-end application, this would be something that is constructed by pulling event data. It would be useful to show that pattern if its not too much to implement.

Reply:
I have added a `generate_proposal_id` and `generate_vote_data`, now there are a _lot_ of caveats, as the signed digest includes a nonce, address of the deployed contract etc. but it's a start for people